config/dynamic.c

```diff
@@ -16,11 +16,467 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <sys/un.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_cycles.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
 #include "gatekeeper_config.h"
+#include "gatekeeper_launch.h"
+
+/* TODO Get the install-path via Makefile. */
+#define LUA_DY_BASE_DIR     "./lua"
+#define DYNAMIC_CONFIG_FILE "dy.lua"
+
+const char *CLIENT_PROC_ERROR =
+	"function reply() return -1, \"Dynamic Configuration cannot process the request. end\"\n";
+
+const uint16_t LUA_MSG_MAX_LEN = 65535;
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, uint16_t reply_len)
+{
+	int send_size;
+	char buffer[reply_len + 2];
+	uint16_t nlen = htons(reply_len);
+
+	/* The first two bytes: the length of the message in network order. */
+	memcpy(buffer, &nlen, sizeof(nlen));
+	memcpy(buffer + sizeof(nlen), reply_msg, reply_len);
+
+	/* Sending the message. */
+	send_size = write(conn_fd, buffer, sizeof(buffer));
+	if (send_size != (int)sizeof(buffer)) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to send reply message (expected len = %hu, sended len = %d): \n %s\n",
+			reply_len, send_size, reply_msg);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+	const char *msg, int msg_len, struct dynamic_config *dy_conf)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	int reply_len;
+	char reply_msg[LUA_MSG_MAX_LEN + 1];
+	memset(reply_msg, 0, sizeof(reply_msg));
+
+	ret = luaL_loadbuffer(dy_conf->lua_state, msg, msg_len, "message") ||
+		lua_pcall(dy_conf->lua_state, 0, 0, 0);
+	if (ret != 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: %s!\n", lua_tostring(dy_conf->lua_state, -1));
+		return -1;
+	}
+
+	/* Function to be called. */
+	lua_getglobal(dy_conf->lua_state, "dy_cmd_proc");
+	lua_pushlightuserdata(dy_conf->lua_state, reply_msg);
+	ret = lua_pcall(dy_conf->lua_state, 1, 1, 0);
```

Comment:
What's the difference between the first chunking calling `dy_cmd_proc()` and the second `lua_pcall()` calling `dy_cmd_proc()`?